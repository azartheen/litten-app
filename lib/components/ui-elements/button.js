/**
 * @format
 * @flow strict-local
 */

import React from 'react'
import { StyleSheet, Text, TouchableOpacity } from 'react-native'
import { vw } from 'react-native-expo-viewport-units'
import colors from 'styles/colors'

const UIButton: () => React$Node = ({
  children,
  disabled = false,
  onPress,
  secondary,
  style,
  textStyle,
  ...otherProps
}) => (
  <TouchableOpacity
    {...otherProps}
    onPress={disabled ? null : onPress}
    style={StyleSheet.compose(
      secondary
        ? disabled
          ? secondaryDisabledStyle
          : secondaryStyle
        : disabled
        ? primaryDisabledStyle
        : primaryStyle,
      style,
    )}>
    <Text style={StyleSheet.compose(styles.uiButtonText, textStyle)}>
      {children}
    </Text>
  </TouchableOpacity>
)

const styles = StyleSheet.create({
  uiButton: {
    height: 40,
    width: vw(77),
    borderRadius: 20,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: colors.purple,
  },
  uiButtonText: {
    color: colors.white,
    fontSize: 16,
    fontWeight: '600',
  },
  secondary: {
    backgroundColor: colors.blue,
  },
  disabled: {
    opacity: 0.3,
  },
})

const primaryStyle = styles.uiButton
const primaryDisabledStyle = StyleSheet.compose(primaryStyle, styles.disabled)
const secondaryStyle = StyleSheet.compose(primaryStyle, styles.secondary)
const secondaryDisabledStyle = StyleSheet.compose(
  secondaryStyle,
  styles.disabled,
)

export default UIButton
