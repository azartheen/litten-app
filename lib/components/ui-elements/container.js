/**
 * @format
 * @flow
 */

import React from 'react'
import { StyleSheet, View } from 'react-native'

const UIContainer: () => React$Node = ({ children, style, ...otherProps }) => (
  <View {...otherProps} style={StyleSheet.compose(styles.uiContainer, style)}>
    {children}
  </View>
)

const styles = StyleSheet.create({
  uiContainer: {
    width: '75%',
  },
})

export default UIContainer
