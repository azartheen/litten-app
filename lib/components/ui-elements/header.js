/**
 * @format
 * @flow
 */

import React from 'react'
import { StyleSheet, Text } from 'react-native'
import colors from 'styles/colors'

const UIHeader: (args: any) => React$Node = ({
  children,
  style,
  subheader = false,
  ...otherProps
}) => (
  <Text
    {...otherProps}
    style={[
      StyleSheet.compose(styles.uiHeader, style),
      subheader ? styles.uiSubHeader : undefined,
    ]}>
    {children}
  </Text>
)

const styles = StyleSheet.create({
  uiHeader: {
    fontSize: 24,
    fontWeight: '600',
    color: colors.black,
    paddingTop: 2,
    paddingBottom: 2,
    marginTop: 2,
    marginBottom: 2,
  },
  uiSubHeader: {
    fontSize: 16,
  },
})

export default UIHeader
