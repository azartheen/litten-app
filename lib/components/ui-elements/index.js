import UIActionSheet from './action-sheet'
import UIButton from './button'
import UIContainer from './container'
import UIDropdown from './dropdown'
import UIHeader from './header'
import UIImagePlaceholder from './image-placeholder'
import UIInput from './input'
import UILink from './link'
import UIListItem from './list-item'
import UILoader from './loader'
import UIMessage from './message'
import UIPasswordInput from './password-input'
import UIProgress from './progress'
import UISearch from './search'
import UISelect from './select'
import UISeparator from './separator'
import UISlider from './slider'
import UISwitch from './switch'
import UIText from './text'
import UITextArea from './textarea'

export {
  UIActionSheet,
  UIButton,
  UIContainer,
  UIDropdown,
  UIHeader,
  UIImagePlaceholder,
  UIInput,
  UILink,
  UIListItem,
  UILoader,
  UIMessage,
  UIPasswordInput,
  UIProgress,
  UISearch,
  UISelect,
  UISeparator,
  UISlider,
  UISwitch,
  UIText,
  UITextArea,
}
