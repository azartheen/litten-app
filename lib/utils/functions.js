/**
 * @format
 * @flow strict-local
 */

import md5Hex from 'md5-hex'
import { translate } from './i18n'

/**
 * Creates an iteratable with 'n' elements starting at 'offset'
 * @param {number} n - Number of elements to create
 * @param {number} [offset=1] - Sets the value for the first element
 * @returns {number[]}
 */
export function iterateTimes(n, offset = 1) {
  if (n < 0) {
    return []
  }
  return [...Array(n)].map((v, i) => i + offset)
}

/**
 * To be used as a callback function to map the countries data source to the
 * UI selector element as options
 * @param {{cca2: string, name: {common: string}}} item - Number of elements to create
 * @returns {{key: string, label: string, value: string}}
 */
export function mapCountriesToSelector(item) {
  if (typeof item === 'object') {
    return {
      key: item.cca2,
      label: item.name?.common,
      value: item.cca2,
    }
  }
}

/**
 * Maps error messages in the translations files by type and error code
 * @param {string} type - Type of the error code to be used as a key prefix
 * @param {string} code - The error code to get the message for
 * @returns {string}
 */
export function getErrorMessage(type, code) {
  const defaultValue = 'Unknown error'
  if (type && typeof type === 'string' && code && typeof code === 'string') {
    const selector = `feedback.${type}ErrorMessages.${code}`
    if (typeof selector === 'string') {
      return translate(selector, { defaultValue })
    }
  }
  return defaultValue
}

/**
 * Parses the user's supplied image URI during the register photo step,
 * if none is supplied, a Gravatar URI is generated based on the user's email
 * @param {(string|null)} avatar - URI location of the user's avatar
 * @param {{email: string}} options - Extra options, e.g. the user's email
 * @returns {string}
 */
export function parseAvatar(avatar = null, { email = '' } = {}) {
  if (typeof avatar === 'string' && avatar.length) {
    return avatar
  }
  // Grab a gravatar
  if (typeof email === 'string' && email.length) {
    // The Gravatar image service
    const gravatarUrl = 'https://s.gravatar.com/avatar'
    // The size query and default if no gravatar exists
    const query = 's=128&d=mp'
    // Gravatar uses MD5 hashes from an email address (all lowercase)
    const hash = md5Hex(email.toLowerCase())
    // The full avatar URL
    avatar = `${gravatarUrl}/${hash}?${query}`
    return avatar
  }
  return null
}

/**
 * Logs whatever is passed as an argument
 * @param {...any} args - List of arguments to be logged
 * @returns {void}
 */
export function logError(...args) {
  if (process.env.NODE_ENV === 'development') {
    console.warn(...args)
  }
}
