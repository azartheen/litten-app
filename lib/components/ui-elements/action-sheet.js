/**
 * @format
 * @flow strict-local
 */

import { ActionSheetIOS, Alert, Platform } from 'react-native'
import { translate } from 'utils/i18n'

const UIActionSheet = (options, callback) => {
  if (Platform.OS === 'ios') {
    ActionSheetIOS.showActionSheetWithOptions(options, callback)
  } else {
    Alert.alert(translate('feedback.errorMessages.notImplemented'))
  }
}

export default UIActionSheet
