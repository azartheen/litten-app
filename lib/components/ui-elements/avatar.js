/**
 * @format
 * @flow
 */

import React from 'react'
import { Image, StyleSheet, View } from 'react-native'
import { USER_AVATAR_SIZE_MINI } from 'utils/constants'

const UIAvatar: (args: any) => React$Node = ({
  containerStyle,
  size,
  source,
  ...otherProps
}) => {
  const getSize = () => {
    switch (size) {
      default:
        return styles.uiAvatarSizeMini
    }
  }

  const getSource = () => {
    if (typeof source === 'object' && source.uri) {
      return {
        ...source,
        cache: 'force-cache',
      }
    }
    return source
  }

  return (
    <View {...otherProps} style={[containerStyle, getSize()]}>
      {source && (
        <Image
          source={getSource()}
          style={[styles.uiAvatarImage, getSize()]}
          resizeMode="contain"
        />
      )}
    </View>
  )
}

const styles = StyleSheet.create({
  uiAvatarImage: {
    overflow: 'hidden',
  },
  uiAvatarSizeMini: {
    height: USER_AVATAR_SIZE_MINI,
    width: USER_AVATAR_SIZE_MINI,
    borderRadius: USER_AVATAR_SIZE_MINI / 2,
  },
})

export default UIAvatar
