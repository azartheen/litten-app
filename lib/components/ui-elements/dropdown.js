/**
 * @format
 * @flow
 */

import React from 'react'
import { Image, StyleSheet, Text, View } from 'react-native'
import {
  Menu,
  MenuOptions,
  MenuOption,
  MenuTrigger,
  renderers,
} from 'react-native-popup-menu'
import { iconChevron } from 'images'
import { rotate } from 'utils/functions'
import colors from 'styles/colors'

const { Popover } = renderers

const UIDropdown: (args: any) => React$Node = ({
  onSelect,
  options,
  placement = 'auto',
  selectedValue,
  style,
  ...otherProps
}) => {
  const translateSelectedValue = () =>
    options.find(({ value }) => value === selectedValue)?.label

  return (
    <View
      {...otherProps}
      style={StyleSheet.compose(styles.uiDropdownContainer, style)}>
      <Menu
        renderer={(props) => (
          <Popover
            {...props}
            placement={placement}
            anchorStyle={styles.anchorStyle}
          />
        )}
        onSelect={(value) => onSelect(value)}>
        <MenuTrigger>
          <View style={styles.uiDropdown}>
            <Text style={styles.selectedValueText}>
              {translateSelectedValue()}
            </Text>
            <Image
              source={iconChevron}
              resizeMode="contain"
              style={styles.iconChevron}
            />
          </View>
        </MenuTrigger>
        <MenuOptions
          customStyles={{ optionsContainer: styles.optionsContainer }}>
          {options.map(({ key, label, value }) => (
            <MenuOption key={key} value={value}>
              <Text
                style={
                  selectedValue === value
                    ? StyleSheet.compose(
                        styles.optionsText,
                        styles.optionsTextActive,
                      )
                    : styles.optionsText
                }>
                {label}
              </Text>
            </MenuOption>
          ))}
        </MenuOptions>
      </Menu>
    </View>
  )
}

const styles = StyleSheet.create({
  uiDropdownContainer: {
    display: 'flex',
    flexDirection: 'row',
    marginTop: 6,
    marginBottom: 6,
  },
  uiDropdown: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
    padding: 12,
    borderRadius: 20,
    backgroundColor: colors.white,
  },
  selectedValueText: {
    fontWeight: '600',
    color: colors.blue,
  },
  iconChevron: {
    maxHeight: 12,
    maxWidth: 12,
    transform: [rotate('90deg', 'Z')],
  },
  optionsContainer: {
    padding: 10,
    borderRadius: 20,
    borderWidth: 0,
    shadowRadius: 0,
    shadowOffset: {
      height: 0,
      width: 0,
    },
    backgroundColor: colors.white,
  },
  optionsText: {
    padding: 2,
    fontWeight: '600',
    color: colors.darkGray,
  },
  optionsTextActive: {
    color: colors.blue,
  },
  anchorStyle: {
    height: 0,
  },
})

export default UIDropdown
