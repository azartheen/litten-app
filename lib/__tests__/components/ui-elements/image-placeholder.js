import React from 'react'
import { UIImagePlaceholder } from 'ui-elements'
import { iterateTimes } from 'utils/functions'
import renderer from 'react-test-renderer'

const allowedNumPhotos = 8
const photos = []

describe('Snapshot test for the "UIImagePlaceholder" component', () => {
  it('renders correctly', () => {
    const element = renderer
      .create(
        <UIImagePlaceholder.Group>
          {iterateTimes(allowedNumPhotos).map((v, idx) =>
            photos[idx] ? (
              <UIImagePlaceholder.ImageItem key={idx} source={photos[idx]} />
            ) : (
              <UIImagePlaceholder.Item
                key={idx}
                actionable={idx === photos.length}
              />
            ),
          )}
        </UIImagePlaceholder.Group>,
      )
      .toJSON()
    expect(element).toMatchSnapshot()
  })
})
