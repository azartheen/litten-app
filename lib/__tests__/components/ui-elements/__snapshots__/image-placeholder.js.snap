// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for the "UIImagePlaceholder" component renders correctly 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "flexWrap": "wrap",
      "justifyContent": "space-between",
      "margin": -3,
    }
  }
>
  <View
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Array [
            Object {
              "aspectRatio": 1,
              "borderColor": "#E5E5E5",
              "borderRadius": 10,
              "borderWidth": 2,
              "height": 157.5,
              "margin": 3,
            },
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
            },
          ],
          Object {
            "backgroundColor": "#39A4F2",
          },
        ]
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "height": 32,
            "margin": 8,
            "tintColor": "#FFFFFF",
            "width": 36,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#FFFFFF",
            "fontSize": 10,
            "fontWeight": "600",
          }
        }
      >
        Add photo
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "borderColor": "#E5E5E5",
          "borderRadius": 10,
          "borderWidth": 2,
          "height": 157.5,
          "margin": 3,
        },
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        },
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;
